src/freenet/client/async/SplitFileInserterStorage.java

```diff
@@ -1760,4 +1762,16 @@ public void clearCooldown() {
         this.callback.clearCooldown();
     }
 
+    /** @return -1 if the insert has finished, 0 if has blocks to send, otherwise Long.MAX_VALUE. */
+    public long getWakeupTime(ClientContext context, long now) {
+        // LOCKING: hasFinished() uses (this), separate from cooldownLock.
+        // It is safe to use both here (on the request selection thread), one after the other.
+        if (hasFinished()) 
+            return -1;
+        if (noBlocksToSend())
+            return Long.MAX_VALUE;
```

Comment:
The function is called inside the lock on the whole request selection tree (in ClientRequestSelector). So in theory thread 2 may set the finished flag, but then it will have to wait for the first thread to complete before removing it from the selection tree.

The bug that I saw was that inserts were finishing and somehow not getting removed from the request selection tree (indicating some other bug / race condition?), and had noBlocksToSend = false (because we were not selecting blocks), so it would return 0, so ClientRequestSelector chooses the request, and it can't create a block to send, so it blocked any other inserts from running.

The simplest solution to that bug is to make getWakeupTime() implement the contract, that is, return -1 when the request has finished already. This does not necessarily indicate an error, so there may be another bug somewhere...

This will be much, much simpler after I have made SplitFileFetcherGet/SplitFileInserterSender be per-segment, which would involve only the segment lock.